find_exponential/src/main.rs

```diff
@@ -0,0 +1,197 @@
+    for _ in 0..num_cpus {
+        tx.send(()).unwrap();
+    }
+    loop {
+        rx.recv().unwrap();
+        let mut literals = literals.clone();
+        let tx = tx.clone();
+        count += BATCH_SIZE as isize;
+        pool.execute(move || {
+            literals.shuffle(&mut rand::thread_rng());
+            let combs = literals.iter()
+                .combinations(COMBINATIONS)
+                .take(BATCH_SIZE);
+            for comb in combs {
+                let res = panic::catch_unwind(|| {
+                    let concatenated = comb.into_iter().flatten().cloned().collect::<Vec<_>>();
+                    let s = String::from_utf8(concatenated).unwrap();
+                    test(&s);
+                });
+                if res.is_err() {
+                    ::std::process::exit(1);
+                }
+            }
+            tx.send(()).unwrap();
+        });
+        if count > 0 {
+            println!("{:<20} throughput: {}", count, count as u64 / start_time.elapsed().as_secs());
+        }
+    }
+}
+
+fn test(pattern: &str) {
+    let (_, _, non_linear) = test_pattern(pattern, SAMPLE_SIZE, false);
+    if non_linear {
+        let (array, score, non_linear) = test_pattern(pattern, SAMPLE_SIZE*2, true);
+        if non_linear {
+            print_find(pattern, &array, score);
+        }
+    }
+}
+
+fn test_pattern(pattern: &str, sample_size: usize, remove_outliers: bool) -> (Array2<f64>, f64, bool) {
+    let s = pattern.repeat(NUM_BYTES / pattern.len());
+    let mut array = Array2::zeros((2, sample_size));
+
+    let mut i = 0;
+    while i < sample_size {
+        let (dur, n) = calculate_point(&s, i+1, sample_size);
+        array[[0, i]] = n as f64;
+        array[[1, i]] = dur.as_nanos() as f64;
+        if DEBUG_LEVEL >= 3 {
+            println!("duration: {}", dur.as_nanos());
+        }
+
+        if remove_outliers && i > 0 && array[[1, i-1]] > array[[1, i]] {
+            // We have an outlier, possibly due to rescheduling.
+            // Redo from the last sample
+            if DEBUG_LEVEL >= 3 {
+                println!("removed outlier");
+            }
+            i -= 1;
+            continue;
+        }
+
+        if dur.as_millis() > MAX_MILLIS {
+            return (array, 0.0, true);
+        }
+        i += 1;
+    }
+
+    let (score, non_linear) = SCORE_FUNCTION(&array);
+
+    if DEBUG_LEVEL >= 1 {
+        println!("{:<30}{}", score, pattern);
+    }
+    if DEBUG_LEVEL >= 2 {
+        println!("{}", array.t());
+    }
+
+    return (array, score, non_linear)
+}
+
+fn calculate_point(s: &str, i: usize, sample_size: usize) -> (Duration, usize) {
+    let mut n = s.len() / sample_size * i;
+    // find closest byte boundary
+    while !s.is_char_boundary(n) {
+        n += 1;
+    }
+    (time_needed(&s[..n]), n)
+}
+
+fn time_needed(s: &str) -> Duration {
+    let parser = Parser::new_ext(&s, Options::all());
+    let time = Instant::now();
+    parser.for_each(|evt| {
```

Comment:
Why is the black box needed? What happens to the event isn't particularly important, right? I think a `parser.count()` would be fine here.

Reply:
I need to use the produced events of the parser. Otherwise the optimizer will realize that the events are unused, meaning the whole parser is unused, removing the whole parsing process. By introducing the black_box, it is guaranteed that the compiler won't optimize out the parser and event generation.

If I use `black_box(parser.count())`, due to `-C lto = fat`, the optimizer *might* be intelligent enough to eliminate `Event` generation and possibly other parts of the parser. That's why instead I opted into black-boxing the events directly, just as any normal user would.

Reply:
Makes sense, good thinking!